### PR TITLE
[train][docs] fix doc search issues, examples gallery & filter

### PR DESCRIPTION
**Pull request by maxpumperla**
[docs preview](https://ray--31635.org.readthedocs.build/en/31635/train/train.html)

Screenshot 2023-01-12 at 17 00 50

### CODE REVIEW

**Comment by maxpumperla** [comment]
@matthewdeng can I get a review from you?!

File: doc/source/_static/js/tags.js
Comment (krfricke) [inline doc/source/_static/js/tags.js:22]: Remove before merge?

**Review by krfricke — COMMENTED** [review]
This is great!

My main confusion is that the "hover" color is almost the same as the "active" color which makes it hard to see if I am activating/deactivating a filter

**Comment by maxpumperla** [comment]
@krfricke thanks, great comments. will address both later

File: doc/source/_toc.yml
Comment (matthewdeng) [inline doc/source/_toc.yml:109]: Can we keep this specifically named `Torch Data Prefetching Benchmark` ?

File: doc/source/train/architecture.rst
Comment (matthewdeng) [inline doc/source/train/architecture.rst:16]: Should we rename this guide to be deep learning specific? The boosted tree trainers actually don't have any of these backend/executor/worker group abstractions.

Also, this diagram is severely outdated 😅

File: doc/source/train/architecture.rst
Comment (matthewdeng) [inline doc/source/train/architecture.rst:29]: Links aren't rendering properly

image

File: doc/source/train/examples.rst
Comment (matthewdeng) [inline doc/source/train/examples.rst:49]: Should this be a 🤗 logo?

**Review by matthewdeng — APPROVED** [review]
This is really nice! I feel like I learned some neat things here - is this in itself something we can codify as best practices?

File: doc/source/train/config_guide.rst
Comment (amogkam) [inline doc/source/train/config_guide.rst:72]: do we need to specify "in Train" for everything? It's already under the Train docs, and makes the headings more verbose.

File: doc/source/train/dl_guide.rst
Comment (amogkam) [inline doc/source/train/dl_guide.rst:447]: ditto here-- do we need to specify "in Ray Train" for all these sections?

File: doc/source/_toc.yml
Comment (amogkam) [inline doc/source/_toc.yml:88]: Boosted Tree may not be well-known parlance.

```suggestion
                title: "XGBoost/LightGBM guide"
```

File: doc/source/conf.py
Comment (amogkam) [inline doc/source/conf.py:363]:
```suggestion
    "trainMLflow": "MLflow,Training",
```

File: doc/source/train/api.rst
Comment (amogkam) [inline doc/source/train/api.rst:179]: The RLTrainer links are no longer working 
image

File: doc/source/train/examples.rst
Comment (amogkam) [inline doc/source/train/examples.rst:10]: When filtering, is it possible to make the headings no longer show up if there are no examples in that category?

**Review by amogkam — COMMENTED** [review]
Thanks @maxpumperla, this looks great! Left some comments

File: doc/source/train/config_guide.rst
Comment (maxpumperla) [reply doc/source/train/config_guide.rst:72]: @amogkam yeah, I understand it looks verbose, but this is to boost doc search results.

Currently, if you search "checkpointing ray train" you just get nothing meaningful back.

File: doc/source/train/examples.rst
Comment (maxpumperla) [reply doc/source/train/examples.rst:10]: It's doable in principle, but requires a bit of JS surgery. I'm taking notes and do this as follow-up.

p.s.: arguably, with a proper tag system you don't need the headings in the first place. many other tools just go for a paginated list of examples, see e.g. https://pytorch.org/tutorials/

File: doc/source/train/api.rst
Comment (maxpumperla) [reply doc/source/train/api.rst:179]: good catch, thanks

File: doc/source/_toc.yml
Comment (maxpumperla) [reply doc/source/_toc.yml:109]: sure! thanks

File: doc/source/train/architecture.rst
Comment (maxpumperla) [reply doc/source/train/architecture.rst:16]: I know... man, this particular doc is painful to read. That's why I touched it up, but it's a bit of a rabbit hole!

File: doc/source/train/architecture.rst
Comment (maxpumperla) [reply doc/source/train/architecture.rst:29]: good catch!

File: doc/source/train/architecture.rst
Comment (maxpumperla) [reply doc/source/train/architecture.rst:16]: I'll add a TODO, but we need to commit to fixing this. I think having a general Architecture Guide is useful to have.

**Comment by maxpumperla** [comment]
@matthewdeng yes, I think we can codify this. I've summarise the approach here: https://docs.google.com/document/d/1cXb1FELZme9zxEYhXXxrGvtLB2xs1auHA46v-1uMYf0/edit#heading=h.8h9qzb62266g

The gallery improvements of this PR are not part of that, but having consistent, self-explaining headers (esp. for examples) is. If we want to boost certain concepts and workloads, ideally we have documents with titles corresponding to these searches.

One example: it's difficult to find good search results (on-site or search engines) on "distributed training with PyTorch/TensorFlow (on Ray)". We cover these workloads in the examples touched in this PR. But if we're literally never mentioning what we're doing anywhere, no paid doc search service can salvage that.

File: doc/source/_static/css/custom.css
Comment (krfricke) [inline doc/source/_static/css/custom.css:491]: Can we limit this lighter color to non-activated buttons? Otherwise when you're on a button it's impossible to see if it's currently activated or not
```suggestion
.tag.btn-outline-primary:hover {
```

File: doc/source/_static/js/tags.js
Comment (krfricke) [inline doc/source/_static/js/tags.js:45]: One thing I noticed is that we start with "all" being selected. If I select e.g. "PyTorch", then "all" gets deselected and "PyTorch" gets selected (Expected). However, when I unselect "PyTorch", "all" does not get selected again. All panels show up so this is more of a detail but maybe it's easy to fix

**Review by krfricke — COMMENTED** [review]
Happy with the content, just two minor comments for the tags

**Review by matthewdeng — APPROVED** [review]

**Review by richardliaw — APPROVED** [review]

**Review by richardliaw — APPROVED** [review]

**Pull request merged**
